src/test/threadpool_tests.cpp

```diff
@@ -0,0 +1,271 @@
+// Copyright (c) The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <util/string.h>
+#include <util/threadpool.h>
+
+#include <boost/test/unit_test.hpp>
+
+BOOST_AUTO_TEST_SUITE(threadpool_tests)
+
+constexpr auto TIMEOUT_SECS = std::chrono::seconds(120);
+
+template <typename T>
+void WaitFor(const std::vector<std::future<T>>& futures, const std::string& context)
+{
+    for (size_t i = 0; i < futures.size(); ++i) {
+        if (futures[i].wait_for(TIMEOUT_SECS) != std::future_status::ready) {
+            throw std::runtime_error("Timeout waiting for: " + context + ", task index " + util::ToString(i));
+        }
+    }
+}
+
+// Block a number of worker threads by submitting tasks that wait on `blocker_future`.
+// Returns the futures of the blocking tasks, ensuring all have started and are waiting.
+std::vector<std::future<void>> BlockWorkers(ThreadPool& threadPool, std::shared_future<void>& blocker_future, int num_of_threads_to_block, const std::string& context)
+{
+    // Per-thread ready promises to ensure all workers are actually blocked
+    std::vector<std::promise<void>> ready_promises(num_of_threads_to_block);
+    std::vector<std::future<void>> ready_futures;
+    ready_futures.reserve(num_of_threads_to_block);
+    for (auto& p : ready_promises) ready_futures.emplace_back(p.get_future());
+
+    // Fill all workers with blocking tasks
+    std::vector<std::future<void>> blocking_tasks;
+    for (int i = 0; i < num_of_threads_to_block; i++) {
+        std::promise<void>& ready = ready_promises[i];
+        blocking_tasks.emplace_back(threadPool.Submit([blocker_future, &ready]() {
+            ready.set_value();
+            blocker_future.wait();
+        }));
+    }
+
+    // Wait until all threads are actually blocked
+    WaitFor(ready_futures, context);
+    return blocking_tasks;
+}
+
+BOOST_AUTO_TEST_CASE(threadpool_basic)
+{
+    // Test Cases
+    // 0) Submit task to a non-started pool.
+    // 1) Submit tasks and verify completion.
+    // 2) Maintain all threads busy except one.
+    // 3) Wait for work to finish.
+    // 4) Wait for result object.
+    // 5) The task throws an exception, catch must be done in the consumer side.
+    // 6) Busy workers, help them by processing tasks from outside.
+    // 7) Recursive submission of tasks.
+    // 8) Submit task when all threads are busy, stop pool and verify the task gets executed.
+
+    const int NUM_WORKERS_DEFAULT = 3;
```

Comment:
I needed at least nproc+1 workers to be able to reliably reproduce concurrency issues (such as the non-atomic `counter` update)
```suggestion
    const size_t NUM_WORKERS_DEFAULT{size_t(GetNumCores()) + 1}; // we need to make sure there's *some* contention
```